### PR TITLE
Allow filtering and completion for machine types

### LINKED ISSUE #1352: Allow filtering and completion for machine types
**What would you like to be added**:
The list of available machine types can be very large and it difficult for the user to find and select the desired machine type. It should be possible to reduce the list by entering filter criteria. 

**Why is this needed**:

### PR DESCRIPTION
**Pull request by holgerkoser**
**What this PR does / why we need it**:
The list of available machine types can be very large and it difficult for the user to find and select the desired machine type. It should be possible to reduce the list by entering filter criteria.

##### Examples:

Only machine types where the name contains  `m5n` 
m5n

Only machine types where the name contains  `m5` and `128Gi` memory 
m5 and 128Gi

Only machine types where the name contains `m5` and `8xl` 
m5 and 8xl

**Which issue(s) this PR fixes**:
Fixes #1352 

**Special notes for your reviewer**:

**Release note**:

```feature user
It is now possible to filter the dropdown list of available machine types by name, cpu, gpu and memory
```

### CODE REVIEW

**Comment by dguendisch** [comment]
Looks good, very much welcomed, thank you!

**Comment by vlerenc** [comment]
That was fast. 🚀 Thank you! 🎉

File: frontend/src/components/ShootWorkers/MachineType.vue [hunk 1]
```diff
@@ -71,62 +63,110 @@ export default {
...
+      }
+      const { name, cpu, gpu, memory, storage } = item
+      const terms = query.split(/\s+/)
+      const properties = [cpu, gpu, memory]
+      if (storage) {
+        properties.push(storage.type)
+        properties.push(storage.class)
+        properties.push(storage.size)
       }
+      return terms.every(term => name?.includes(term) || properties.includes(term))
```
Comment (grolu) [inline frontend/src/components/ShootWorkers/MachineType.vue:146]:
```suggestion
      return terms.every(term => name?.includes(term) || some(properties, prop => prop?.includes(term)))
```
I would change it to something like this to allow partial match also for the properties. Moreover, you could do a `toLowerCase()` for all the strings before comparing them.

File: frontend/src/components/ShootWorkers/MachineType.vue [hunk 2]
```diff
@@ -71,62 +63,110 @@ export default {
...
-        this.$emit('valid', { id: this.worker.id, valid: this.valid })
+    filter (item, query) {
+      if (!item) {
+        return false
+      }
+      if (typeof query !== 'string') {
+        return true
+      }
+      const { name, cpu, gpu, memory, storage } = item
+      const terms = query.split(/\s+/)
```
Comment (petersutter) [inline frontend/src/components/ShootWorkers/MachineType.vue:139]: is there a difference to `query.split(" ")`?

Comment (holgerkoser) [reply frontend/src/components/ShootWorkers/MachineType.vue:139]: > is there a difference to `query.split(" ")`?

image

**Review by petersutter — APPROVED** [review]
/lgtm

File: frontend/src/components/ShootWorkers/MachineType.vue (hunk 2 above)
Comment (holgerkoser) [reply frontend/src/components/ShootWorkers/MachineType.vue:139]: May be we should do:
```js
'   a\tb  c\nd\re   '.replace(/(^\s+|\s+$)/g, '').split(/\s+/)
```

File: frontend/src/components/ShootWorkers/MachineType.vue (hunk 1 above)
Comment (holgerkoser) [reply frontend/src/components/ShootWorkers/MachineType.vue:155]: ✔️

File: frontend/src/components/ShootWorkers/MachineType.vue (hunk 2 above)
Comment (holgerkoser) [reply frontend/src/components/ShootWorkers/MachineType.vue]: ✔️

**Pull request merged**